rasa/utils/tensorflow/model_data_utils.py

```diff
@@ -1,226 +1,370 @@
-            ACTION_TEXT, ENTITIES, SLOTS, FORM) to a list of features for all
-            dialogue turns in all training trackers
-        zero_state_features: Contains default feature values for attributes
+        features: a dictionary of attributes to a list of features for all
+            examples in the training data
+        fake_features: Contains default feature values for attributes
+        consider_dialogue_dimension: If set to false the dialogue dimension will be
+            removed from the resulting sequence features.
+        featurizers: the featurizers to consider
 
     Returns:
-        Input in "Data" format and zero state features
+        Input in "Data" format and zero features
     """
     training = False
-    if not zero_state_features:
+    if not fake_features:
         training = True
-        zero_state_features = defaultdict(list)
+        fake_features = defaultdict(list)
 
     # unify format of incoming features
-    if isinstance(tracker_state_features[0], Dict):
-        tracker_state_features = [[dicts] for dicts in tracker_state_features]
+    if isinstance(features[0], Dict):
+        features = [[dicts] for dicts in features]
 
-    state_to_tracker_features = surface_attributes(tracker_state_features)
+    attribute_to_features = _surface_attributes(features, featurizers)
 
     attribute_data = {}
 
     # During prediction we need to iterate over the zero features attributes to
     # have all keys in the resulting model data
     if training:
-        attributes = list(state_to_tracker_features.keys())
+        attributes = list(attribute_to_features.keys())
     else:
-        attributes = list(zero_state_features.keys())
+        attributes = list(fake_features.keys())
 
     # In case an attribute is not present during prediction, replace it with
     # None values that will then be replaced by zero features
     dialogue_length = 1
-    for tracker_features in state_to_tracker_features.values():
-        dialogue_length = max(dialogue_length, len(tracker_features[0]))
-    empty_features = [[None] * dialogue_length]
+    num_examples = 1
+    for _features in attribute_to_features.values():
+        num_examples = max(num_examples, len(_features))
+        dialogue_length = max(dialogue_length, len(_features[0]))
+    empty_features = [[None] * dialogue_length] * num_examples
 
     for attribute in attributes:
         attribute_data[attribute] = _features_for_attribute(
             attribute,
             empty_features,
-            state_to_tracker_features,
+            attribute_to_features,
             training,
-            zero_state_features,
+            fake_features,
+            consider_dialogue_dimension,
         )
 
     # ensure that all attributes are in the same order
     attribute_data = OrderedDict(sorted(attribute_data.items()))
 
-    return attribute_data, zero_state_features
+    return attribute_data, fake_features
 
 
 def _features_for_attribute(

```

Comment:
Optional: Rename this function to `_feature_arrays_for_attribute` or similar. There are so many places where we are collecting `features` for attributes that it gets confusing when you reach this function.

Reply:
@tabergma could you please take this over